### ISSUE TITLE
Dart and analyzer produce different types of compile errors when check least and greatest closures for FutureOr<X>.

### ISSUE THREAD

**Issue by iarkh** [dart-lang/sdk#44070]
Dart SDK version: 2.12.0-13.0.dev (dev) (Mon Nov 2 15:57:37 2020 -0800) on "windows_x64"

Here is a source code example which demonstrates the problem:
```dart
import "dart:async";

Type? _capturedTypeArgument;

X captureTypeArgument<X>() {
  _capturedTypeArgument = X;
  throw "Error";
}

Type? get capturedTypeArgument {
  var result = _capturedTypeArgument;
  _capturedTypeArgument = null;
  return result;
}

void main() {
  void f<X>(FutureOr<X> Function() g) => g();

  f(() => captureTypeArgument()..unknownMember);
  f(() => captureTypeArgument()?..unknownMember);
}
```
Both analyzer and dart produce a compile error here.

The problem is that these are different types of errors for different pieсes of the failing lines.

Seems like both tools should complain about the same here.

Sample output is:

> $> dartanalyzer test.dart
> Analyzing test.dart...
>   error - An expression whose value can be 'null' must be null-checked before it can be dereferenced. - test.dart:19:11 - unchecked_use_of_nullable_value
>   error - An expression whose value can be 'null' must be null-checked before it can be dereferenced. - test.dart:20:11 - unchecked_use_of_nullable_value
> 2 errors found.
> 
> $> dart test.dart
>  test.dart:19:34: Error: The getter 'unknownMember' isn't defined for the class 'FutureOr<Object?>?'.
>  - 'Object' is from 'dart:core'.
> Try correcting the name to the name of an existing getter, or defining a getter or field named 'unknownMember'.
>   f(() => captureTypeArgument()..unknownMember);
>                                  ^^^^^^^^^^^^^
> test.dart:20:35: Error: The getter 'unknownMember' isn't defined for the class 'FutureOr<Object?>'.
>  - 'Object' is from 'dart:core'.
> Try correcting the name to the name of an existing getter, or defining a getter or field named 'unknownMember'.
>   f(() => captureTypeArgument()?..unknownMember);
>                                   ^^^^^^^^^^^^^

Please also note that the following code statically passes with both tools, no `An expression whose value can be 'null' must be null-checked before it can be dereferenced`  error message here:
``` dart
  f(() => captureTypeArgument()..toString());
  f(() => captureTypeArgument()?..toString());
```

**Comment by eernstg** [comment]
It looks like the error from the analyzer at line 20 is spurious: `?..` is allowed on a value whose type is potentially nullable.

It is working as intended that `toString()` is allowed: even `null` has the members of `Object`.

The error message from the front end indicating that the type of the receiver is `FutureOr<Object?>?` is misleading: The greatest closure of `FutureOr<U>` with respect to some set of type variables `L` is `FutureOr<V>`, where `V` is the greatest closure of `U` with respect to `L`. This might be caused by the semantic representation of nullability in the front end (the outermost `?` makes no difference), but it is still somewhat confusing.

It could be argued that the two error messages are about the same thing after all: The analyzer says that a potentially nullable expression does not have any members, and the CFE says that the given receiver type doesn't have the specific member which is requested.

However, considering that we allow `null.toString()` it isn't actually true that

> An expression whose value can be 'null' must 
> be null-checked before it can be dereferenced

The trade-off is that the CFE uses an error message which does not emphasize the fact that this is about null.

@stereotype441, @johnniwinther, do you see an useful improvement in making these error messages more similar to each other? In that case we may create new issues about this change, and possibly about the other things mentioned above.

**Comment by stereotype441** [comment]
Yeah, I consider this an analyzer bug.  Although all the specification demands is that the analyzer produce at least one error for erroneous code, the error it produces is *incredibly* confusing--it says you should null check the value, even though you're already null checking it!

Reading between the lines, I suspect that what's happening is that the analyzer checks expressions of the form `x?..m` by converting the type of `x` to its non-nullable counterpart and then proceeding with the logic that would be otherwise used for `x..m`.  But that's not a good way to implement it because types of the form `FutureOr<T>` (where `T` is nullable) don't have a non-nullable counterpart.

The problem affects any erroneous property access or method call applied to `FutureOr` of a nullable type, e.g.:
```dart
import 'dart:async';

main() {
  FutureOr<int?> f;
  f?.m;
  f?.m();
  f?..m;
  f?..m();
}
```

Not super-high priority to fix, since both implementations are correctly rejecting the code, but it would be really nice for the analyzer error message to look more like the front end error message for these cases.

**Comment by bwilkerson** [comment]
I agree. Given code of the form
```
void f(String? s) {
    s.unknown;
}
```
The message ought to indicate that the getter `unknown` isn't defined for the type `String` (or technically `String?`, though the nullability isn't the issue), and the front end's message does that while analyzer's fails to do that.

But given code of the form
```
void f(String? s) {
    s.length;
}
```
It would be better to indicate that `s` needs to be null checked because that's more likely to suggest the right fix to the user. The front end's message in this case ("Property 'length' cannot be accessed on 'String?' because it is potentially null.") is fine. I just don't want analyzer to start producing the "isn't defined" message in this latter case, and because it currently produces the same message in both cases, I think that's a possibility if the implementer isn't careful.

**Comment by srawlins** [comment]
Analyzer behaves the same today for all of the examples.
